src/lib/storage/dictionary_segment.cpp

```diff
@@ -0,0 +1,126 @@
+#include "dictionary_segment.hpp"
+#include <algorithm>
+#include <set>
+#include "fixed_width_attribute_vector.hpp"
+#include "resolve_type.hpp"
+#include "type_cast.hpp"
+#include "utils/assert.hpp"
+#include "value_segment.hpp"
+
+namespace opossum {
+
+template <typename T>
+DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
+  DebugAssert(abstract_segment->size() > 0, "Input segment must contain values.");
+
+  // For now, we can assume to only receive a ValueSegment
+  const auto value_segment = std::static_pointer_cast<ValueSegment<T>>(abstract_segment);
+  const auto values = value_segment->values();
+  std::set<T> distinct_values(values.begin(), values.end());
+
+  // Populate the _dictionary with the unique values.
+  _dictionary.reserve(distinct_values.size());
+  std::copy(distinct_values.begin(), distinct_values.end(), std::back_inserter(_dictionary));
```

Comment:
Good idea to use a set, so sorting and duplicate removal is automated.

Reply:
Thanks!